# Predicting movie popularity using IMDB movies dataset

In [92]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [93]:
# importing the dataset
data = pd.read_csv('movie_metadata.csv')

In [94]:
data

color       director_name  num_critic_for_reviews  duration  \
0                Color       James Cameron                   723.0     178.0   
1                Color      Gore Verbinski                   302.0     169.0   
2                Color          Sam Mendes                   602.0     148.0   
3                Color   Christopher Nolan                   813.0     164.0   
4                  NaN         Doug Walker                     NaN       NaN   
5                Color      Andrew Stanton                   462.0     132.0   
6                Color           Sam Raimi                   392.0     156.0   
7                Color        Nathan Greno                   324.0     100.0   
8                Color         Joss Whedon                   635.0     141.0   
9                Color         David Yates                   375.0     153.0   
10               Color         Zack Snyder                   673.0     183.0   
11               Color        Bryan Singer                   434.0     169.0   
12               Color        Marc Forster                   403.0     106.0   
13               Color      Gore Verbinski                   313.0     151.0   
14               Color      Gore Verbinski                   450.0     150.0   
15               Color         Zack Snyder                   733.0     143.0   
16               Color      Andrew Adamson                   258.0     150.0   
17               Color         Joss Whedon                   703.0     173.0   
18               Color        Rob Marshall                   448.0     136.0   
19               Color    Barry Sonnenfeld                   451.0     106.0   
20               Color       Peter Jackson                   422.0     164.0   
21               Color           Marc Webb                   599.0     153.0   
22               Color        Ridley Scott                   343.0     156.0   
23               Color       Peter Jackson                   509.0     186.0   
24               Color         Chris Weitz                   251.0     113.0   
25               Color       Peter Jackson                   446.0     201.0   
26               Color       James Cameron                   315.0     194.0   
27               Color       Anthony Russo                   516.0     147.0   
28               Color          Peter Berg                   377.0     131.0   
29               Color     Colin Trevorrow                   644.0     124.0   
...                ...                 ...                     ...       ...   
5013             Color          Eric Eason                    28.0      79.0   
5014             Color            Uwe Boll                    58.0      80.0   
5015   Black and White   Richard Linklater                    61.0     100.0   
5016             Color     Joseph Mazzella                     NaN      90.0   
5017             Color        Travis Legge                     1.0      90.0   
5018             Color       Alex Kendrick                     5.0     120.0   
5019             Color       Marcus Nispel                    43.0      91.0   
5020               NaN     Brandon Landers                     NaN     143.0   
5021             Color         Jay Duplass                    51.0      85.0   
5022   Black and White          Jim Chuchu                     6.0      60.0   
5023             Color          Daryl Wein                    22.0      88.0   
5024             Color         Jason Trost                    42.0      78.0   
5025             Color         John Waters                    73.0     108.0   
5026             Color     Olivier Assayas                    81.0     110.0   
5027             Color        Jafar Panahi                    64.0      90.0   
5028   Black and White       Ivan Kavanagh                    12.0      83.0   
5029             Color    Kiyoshi Kurosawa                    78.0     111.0   
5030             Color        Tadeo Garcia                     NaN      84.0   
5031             

## Data exploration

In [95]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5043 entries, 0 to 5042
Data columns (total 28 columns):
color                        5024 non-null object
director_name                4939 non-null object
num_critic_for_reviews       4993 non-null float64
duration                     5028 non-null float64
director_facebook_likes      4939 non-null float64
actor_3_facebook_likes       5020 non-null float64
actor_2_name                 5030 non-null object
actor_1_facebook_likes       5036 non-null float64
gross                        4159 non-null float64
genres                       5043 non-null object
actor_1_name                 5036 non-null object
movie_title                  5043 non-null object
num_voted_users              5043 non-null int64
cast_total_facebook_likes    5043 non-null int64
actor_3_name                 5020 non-null object
facenumber_in_poster         5030 non-null float64
plot_keywords                4890 non-null object
movie_imdb_link              5043 non-

In [96]:
# number of unique values in each column
data.nunique()

color                           2
director_name                2398
num_critic_for_reviews        528
duration                      191
director_facebook_likes       435
actor_3_facebook_likes        906
actor_2_name                 3032
actor_1_facebook_likes        878
gross                        4035
genres                        914
actor_1_name                 2097
movie_title                  4917
num_voted_users              4826
cast_total_facebook_likes    3978
actor_3_name                 3521
facenumber_in_poster           19
plot_keywords                4760
movie_imdb_link              4919
num_user_for_reviews          954
language                       47
country                        65
content_rating                 18
budget                        439
title_year                     91
actor_2_facebook_likes        917
imdb_score                     78
aspect_ratio                   22
movie_facebook_likes          876
dtype: int64

In [97]:
# See the numeric attributes of the dataset
data.describe(include = [np.number])

num_critic_for_reviews     duration  director_facebook_likes  \
count             4993.000000  5028.000000              4939.000000   
mean               140.194272   107.201074               686.509212   
std                121.601675    25.197441              2813.328607   
min                  1.000000     7.000000                 0.000000   
25%                 50.000000    93.000000                 7.000000   
50%                110.000000   103.000000                49.000000   
75%                195.000000   118.000000               194.500000   
max                813.000000   511.000000             23000.000000   

       actor_3_facebook_likes  actor_1_facebook_likes         gross  \
count             5020.000000             5036.000000  4.159000e+03   
mean               645.009761             6560.047061  4.846841e+07   
std               1665.041728            15020.759120  6.845299e+07   
min                  0.000000                0.000000  1.620000e+02   
25%                133.000000              614.000000  5.340988e+06   
50%                371.500000              988.000000  2.551750e+07   
75%                636.000000            11000.000000  6.230944e+07   
max              23000.000000           640000.000000  7.605058e+08   

       num_voted_users  cast_total_facebook_likes  facenumber_in_poster  \
count     5.043000e+03                5043.000000           5030.000000   
mean      8.366816e+04                9699.063851              1.371173   
std       1.384853e+05               18163.799124              2.013576   
min       5.000000e+00                   0.000000              0.000000   
25%       8.593500e+03                1411.000000              0.000000   
50%       3.435900e+04                3090.000000              1.000000   
75%       9.630900e+04               13756.500000              2.000000   
max       1.689764e+06              656730.000000             43.000000   

       num_user_for_reviews        budget   title_year  \
count           5022.000000  4.551000e+03  4935.000000   
mean             272.770808  3.975262e+07  2002.470517   
std              377.982886  2.061149e+08    12.474599   
min                1.000000  2.180000e+02  1916.000000   
25%               65.000000  6.000000e+06  1999.000000   
50%              156.000000  2.000000e+07  2005.000000   
75%              326.000000  4.500000e+07  2011.000000   
max             5060.000000  1.221550e+10  2016.000000   

       actor_2_facebook_likes   imdb_score  aspect_ratio  movie_facebook_likes  
count             5030.000000  5043.000000   4714.000000           5043.000000  
mean              1651.754473     6.442138      2.220403           7525.964505  
std               4042.438863     1.125116      1.385113          19320.445110  
min                  0.000000     1.600000      1.180000              0.000000  
25%                281.000000     5.800000      1.850000              0.000000  
50%                595.000000     6.600000      2.350000            166.000000  
75%                918.000000     7.200000      2.350000           3000.000000  
max             137000.000000     9.500000     16.000000         349000.000000

In [98]:
# Number of unique categorical attributes of the dataset
data.describe(include = [np.object])

color     director_name    actor_2_name genres    actor_1_name  \
count    5024              4939            5030   5043            5036   
unique      2              2398            3032    914            2097   
top     Color  Steven Spielberg  Morgan Freeman  Drama  Robert De Niro   
freq     4815                26              20    236              49   

                 movie_title actor_3_name   plot_keywords  \
count                   5043         5020            4890   
unique                  4917         3521            4760   
top     Victor Frankenstein    John Heard  based on novel   
freq                       3            8               4   

                                          movie_imdb_link language country  \
count                                                5043     5031    5038   
unique                                               4919       47      65   
top     http://www.imdb.com/title/tt0077651/?ref_=fn_t...  English     USA   
freq                                                    3     4704    3807   

       content_rating  
count            4740  
unique             18  
top                 R  
freq             2118

## Data preparation

### Removing duplicates

First we start off by checking for duplicates and removing them.

In [99]:
# checking for duplicate rows based on movie title
print('entries from dataset:', data.shape[0])
duplicates = data.duplicated()
print('number of duplicated entries:', sum(duplicates))

entries from dataset: 5043
number of duplicated entries: 45


In [100]:
# removing the duplicates
data = data.drop_duplicates()
print('entries without duplicates:', data.shape[0])

entries without duplicates: 4998


### Removing features that won't help us determine movie success

Is the color impactful to the model?

In [101]:
data['color'].describe()

count      4979
unique        2
top       Color
freq       4772
Name: color, dtype: object

We can see that out of the 4979 entries for "color", 4772 were "Color". Nearly 96% of movies are colored. We remove this feature.

In [102]:
data = data.drop(['color'], axis = 1)

The IMDB links are also useless to us so let's remove those as well.

In [103]:
data = data.drop(['movie_imdb_link'], axis = 1)

In [104]:
data

director_name  num_critic_for_reviews  duration  \
0          James Cameron                   723.0     178.0   
1         Gore Verbinski                   302.0     169.0   
2             Sam Mendes                   602.0     148.0   
3      Christopher Nolan                   813.0     164.0   
4            Doug Walker                     NaN       NaN   
5         Andrew Stanton                   462.0     132.0   
6              Sam Raimi                   392.0     156.0   
7           Nathan Greno                   324.0     100.0   
8            Joss Whedon                   635.0     141.0   
9            David Yates                   375.0     153.0   
10           Zack Snyder                   673.0     183.0   
11          Bryan Singer                   434.0     169.0   
12          Marc Forster                   403.0     106.0   
13        Gore Verbinski                   313.0     151.0   
14        Gore Verbinski                   450.0     150.0   
15           Zack Snyder                   733.0     143.0   
16        Andrew Adamson                   258.0     150.0   
17           Joss Whedon                   703.0     173.0   
18          Rob Marshall                   448.0     136.0   
19      Barry Sonnenfeld                   451.0     106.0   
20         Peter Jackson                   422.0     164.0   
21             Marc Webb                   599.0     153.0   
22          Ridley Scott                   343.0     156.0   
23         Peter Jackson                   509.0     186.0   
24           Chris Weitz                   251.0     113.0   
25         Peter Jackson                   446.0     201.0   
26         James Cameron                   315.0     194.0   
27         Anthony Russo                   516.0     147.0   
28            Peter Berg                   377.0     131.0   
29       Colin Trevorrow                   644.0     124.0   
...                  ...                     ...       ...   
5013          Eric Eason                    28.0      79.0   
5014            Uwe Boll                    58.0      80.0   
5015   Richard Linklater                    61.0     100.0   
5016     Joseph Mazzella                     NaN      90.0   
5017        Travis Legge                     1.0      90.0   
5018       Alex Kendrick                     5.0     120.0   
5019       Marcus Nispel                    43.0      91.0   
5020     Brandon Landers                     NaN     143.0   
5021         Jay Duplass                    51.0      85.0   
5022          Jim Chuchu                     6.0      60.0   
5023          Daryl Wein                    22.0      88.0   
5024         Jason Trost                    42.0      78.0   
5025         John Waters                    73.0     108.0   
5026     Olivier Assayas                    81.0     110.0   
5027        Jafar Panahi                    64.0      90.0   
5028       Ivan Kavanagh                    12.0      83.0   
5029    Kiyoshi Kurosawa                    78.0     111.0   
5030        Tadeo Garcia                     NaN      84.0   
5031  Thomas L. Phillips                    13.0      82.0   
5032     Ash Baron-Cohen                    10.0      98.0   
5033       Shane Carruth                   143.0      77.0   
5034    Neill Dela Llana                    35.0      80.0   
5035    Robert Rodriguez                    56.0      81.0   
5036     Anthony Vallone                     NaN      84.0   
5037        Edward Burns                    14.0      95.0   
5038         Scott Smith                     1.0      87.0   
5039                 NaN                    43.0      43.0   
5040    Benjamin Roberds                    13.0      76.0   
5041         Daniel Hsia                    14.0     100.0   
5042            Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes          actor_2_name  \
0                         0.0                   855.0      Joel David Moore   
1       

Another column we will remove is the 'plot keywords'. There is just too much diversity in this feature for it to be a good predictor of the score and revenue.

In [107]:
data = data.drop(['plot_keywords'], axis = 1)
data

director_name  num_critic_for_reviews  duration  \
0          James Cameron                   723.0     178.0   
1         Gore Verbinski                   302.0     169.0   
2             Sam Mendes                   602.0     148.0   
3      Christopher Nolan                   813.0     164.0   
4            Doug Walker                     NaN       NaN   
5         Andrew Stanton                   462.0     132.0   
6              Sam Raimi                   392.0     156.0   
7           Nathan Greno                   324.0     100.0   
8            Joss Whedon                   635.0     141.0   
9            David Yates                   375.0     153.0   
10           Zack Snyder                   673.0     183.0   
11          Bryan Singer                   434.0     169.0   
12          Marc Forster                   403.0     106.0   
13        Gore Verbinski                   313.0     151.0   
14        Gore Verbinski                   450.0     150.0   
15           Zack Snyder                   733.0     143.0   
16        Andrew Adamson                   258.0     150.0   
17           Joss Whedon                   703.0     173.0   
18          Rob Marshall                   448.0     136.0   
19      Barry Sonnenfeld                   451.0     106.0   
20         Peter Jackson                   422.0     164.0   
21             Marc Webb                   599.0     153.0   
22          Ridley Scott                   343.0     156.0   
23         Peter Jackson                   509.0     186.0   
24           Chris Weitz                   251.0     113.0   
25         Peter Jackson                   446.0     201.0   
26         James Cameron                   315.0     194.0   
27         Anthony Russo                   516.0     147.0   
28            Peter Berg                   377.0     131.0   
29       Colin Trevorrow                   644.0     124.0   
...                  ...                     ...       ...   
5013          Eric Eason                    28.0      79.0   
5014            Uwe Boll                    58.0      80.0   
5015   Richard Linklater                    61.0     100.0   
5016     Joseph Mazzella                     NaN      90.0   
5017        Travis Legge                     1.0      90.0   
5018       Alex Kendrick                     5.0     120.0   
5019       Marcus Nispel                    43.0      91.0   
5020     Brandon Landers                     NaN     143.0   
5021         Jay Duplass                    51.0      85.0   
5022          Jim Chuchu                     6.0      60.0   
5023          Daryl Wein                    22.0      88.0   
5024         Jason Trost                    42.0      78.0   
5025         John Waters                    73.0     108.0   
5026     Olivier Assayas                    81.0     110.0   
5027        Jafar Panahi                    64.0      90.0   
5028       Ivan Kavanagh                    12.0      83.0   
5029    Kiyoshi Kurosawa                    78.0     111.0   
5030        Tadeo Garcia                     NaN      84.0   
5031  Thomas L. Phillips                    13.0      82.0   
5032     Ash Baron-Cohen                    10.0      98.0   
5033       Shane Carruth                   143.0      77.0   
5034    Neill Dela Llana                    35.0      80.0   
5035    Robert Rodriguez                    56.0      81.0   
5036     Anthony Vallone                     NaN      84.0   
5037        Edward Burns                    14.0      95.0   
5038         Scott Smith                     1.0      87.0   
5039                 NaN                    43.0      43.0   
5040    Benjamin Roberds                    13.0      76.0   
5041         Daniel Hsia                    14.0     100.0   
5042            Jon Gunn                    43.0      90.0   

      director_facebook_likes  actor_3_facebook_likes          actor_2_name  \
0                         0.0                   855.0      Joel David Moore   
1       

## Data Cleaning

We need to clean up NaN and null values.

In [111]:
nandata = data.isnull().sum().to_frame('Number of NaN').sort_values(by = ['Number of NaN'], ascending = False)
nandata

Number of NaN
gross                                874
budget                               487
aspect_ratio                         327
content_rating                       301
title_year                           107
director_name                        103
director_facebook_likes              103
num_critic_for_reviews                49
actor_3_facebook_likes                23
actor_3_name                          23
num_user_for_reviews                  21
duration                              15
actor_2_name                          13
actor_2_facebook_likes                13
facenumber_in_poster                  13
language                              12
actor_1_facebook_likes                 7
actor_1_name                           7
country                                5
imdb_score                             0
cast_total_facebook_likes              0
num_voted_users                        0
movie_title                            0
genres                                 0
movie_facebook_likes                   0

We see that gross and budget have the most NaN values. These are two of the most important predictors so we want to keep them. In order to do so, we must remove the rows that have missing values.